Change Data Capture

lhirata opened the pull request:
Closes #2993.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/3308)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a8a33012e6cbf96776e4b8e499bc0777dc32e96a/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/78be40f91c6e33c513af7adb4178e8c80604bebc/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8d4fad12ed9cbb675a57eebbdc3caf0dd4fda240/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a004b8a8b4bbbc580bb84ccc125ff7ef54172e52/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/49eb216123131a53437ff7bad685ab12c1b56edd/

lhirata commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/create-changefeed.md, line 25 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6wQ8L4Vu4rOWmzwID:-LG6wQ8L4Vu4rOWmzwIE:b-d2iidj) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L25)):*
> ```Markdown
> ## Synopsis
> 
> {% include sql/{{ page.version.version }}/diagrams/create_database.html %}
> ```

TO DO: generate new diagram

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Looking good so far! Here's my first pass. I'll think some more about the ordering guarantees section and also get you the known limitations

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 3 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6rwOf6GPD7Rdvivs9:-LG6rwOf6GPD7RdvivsA:b-m3ge2e) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L3)):*
> ```Markdown
> ---
> title: Change Data Capture (CDC)
> summary: Partitioning is an enterprise feature that gives you row-level control of how and where your data is stored.
> ```

copy paste error?

---

*[v2.1/change-data-capture.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG70sXm8MfuZjwy6X1D:-LG70sXm8MfuZjwy6X1E:b-3phqlm) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L53)):*
> ```Markdown
> For more information, see [`CREATE CHANGEFEED`](create-changefeed.html).
> 
> ### Pause
> ```

also unpause I guess?

---

*[v2.1/change-data-capture.md, line 158 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG70yol4e3Yj9yvMCo7:-LG70yol4e3Yj9yvMCo8:bc3biha) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L158)):*
> ```Markdown
>     This will start up the changefeed in the background and return the `job_id`. The changefeed writes to Kafka and survives node failures.
> 
> 9. In a new terminal, start watching the Kafka topic:
> ```

we should note that the initial scan has output the state of the table as of when the changefeed started (so the initial value of "Petee" is missing")

---

*[v2.1/create-changefeed.md, line 29 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6xa7w08rN2YAAGUEi:-LG6xa7w08rN2YAAGUEj:bh16ams) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L29)):*
> ```Markdown
> ## Parameters
> 
> Parameter | Description
> ```

we should also document `INTO` and `WITH` (https://github.com/cockroachdb/cockroach/blob/fb4e761f83623e862d9896af7cded1f7dc5770e4/pkg/ccl/changefeedccl/changefeed_stmt.go#L36-45)

the `INTO` uri is very similar to what we've done with BACKUP/RESTORE/IMPORT where the "scheme" of the uri indicates what type it is and there are query parameters that vary per type.

- the `kafka` scheme currently only has `topic_prefix` which adds a prefix to all the topic names it has
So `CREATE CHANGEFEED FOR TABLE foo INTO 'kafka://...?topic_prefix=bar_'` would emit things under topic `bar_foo` instead of `foo`.

the available WITH parameters are  `cursor`, `envelope`, and `timestamps`

- `WITH timestamps` turns on the updated and resolved timestamps that we talk about in the ordering section. we might make this the default in 2.1
- `WITH envelope=key_only` only emits the key and no value (this is faster and some people only want to know when the key changes). `WITH envelope=row` is the default. I'm fine leaving this out of the alpha docs if you want. in 2.1, there will also be a `WITH envelope= diff` that emits the old and new value of the changed row, but this is not in the alpha
- `WITH cursor=<timestamp>` starts emitting any changes after the given timestamp, but doesn't output the current state of the table first. if `cursor` is not specified, the changefeed starts by doing a consistent scan of all the watched rows and emits the current value, then moves to emitting any changes that happen after that
- `WITH format=json` is the default. in 2.1, we'll also have `WITH format=avro`

---

*[v2.1/create-changefeed.md, line 32 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6vdaSEha0LeK_15Vl:-LG6vdaSEha0LeK_15Vm:b-fiz3h8) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L32)):*
> ```Markdown
> ----------|------------
> `IF NOT EXISTS` | Create a new changefeed only if a changefeed of the same name does not already exist; if one does exist, do not return an error.
> `name` | The name of the changefeed to create, which [must be unique](#create-fails-name-already-in-use) and follow these [identifier rules](keywords-and-identifiers.html#identifiers).
> ```

note: name will be in 2.1 but is not present on the alpha. I hadn't thought about IF NOT EXISTS, but that's a good idea. also not in the alpha

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308#-:-LG71k1M8xnE5hQntmnX:bg9mfm9)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 3 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6rwOf6GPD7Rdvivs9:-LG7JT7hAB7Lh65FXnXF:bxtsv4q) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L3)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

copy paste error?
</blockquote></details>

oops. fixed

---

*[v2.1/change-data-capture.md, line 53 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG70sXm8MfuZjwy6X1D:-LG7KMlhE_Z7vQ8lxQwV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L53)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

also unpause I guess?
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 158 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG70yol4e3Yj9yvMCo7:-LG7L9KX4qCJOP-Ilq6r:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/change-data-capture.md#L158)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we should note that the initial scan has output the state of the table as of when the changefeed started (so the initial value of "Petee" is missing")
</blockquote></details>

Done.

---

*[v2.1/create-changefeed.md, line 29 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6xa7w08rN2YAAGUEi:-LGGTnr53mi51uMKcJZ1:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L29)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we should also document `INTO` and `WITH` (https://github.com/cockroachdb/cockroach/blob/fb4e761f83623e862d9896af7cded1f7dc5770e4/pkg/ccl/changefeedccl/changefeed_stmt.go#L36-45)

the `INTO` uri is very similar to what we've done with BACKUP/RESTORE/IMPORT where the "scheme" of the uri indicates what type it is and there are query parameters that vary per type.

- the `kafka` scheme currently only has `topic_prefix` which adds a prefix to all the topic names it has
So `CREATE CHANGEFEED FOR TABLE foo INTO 'kafka://...?topic_prefix=bar_'` would emit things under topic `bar_foo` instead of `foo`.

the available WITH parameters are  `cursor`, `envelope`, and `timestamps`

- `WITH timestamps` turns on the updated and resolved timestamps that we talk about in the ordering section. we might make this the default in 2.1
- `WITH envelope=key_only` only emits the key and no value (this is faster and some people only want to know when the key changes). `WITH envelope=row` is the default. I'm fine leaving this out of the alpha docs if you want. in 2.1, there will also be a `WITH envelope= diff` that emits the old and new value of the changed row, but this is not in the alpha
- `WITH cursor=<timestamp>` starts emitting any changes after the given timestamp, but doesn't output the current state of the table first. if `cursor` is not specified, the changefeed starts by doing a consistent scan of all the watched rows and emits the current value, then moves to emitting any changes that happen after that
- `WITH format=json` is the default. in 2.1, we'll also have `WITH format=avro`
</blockquote></details>

Done.

---

*[v2.1/create-changefeed.md, line 32 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6vdaSEha0LeK_15Vl:-LG7Lrft1aujzrGT_eVq:blgqq0l) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L32)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

note: name will be in 2.1 but is not present on the alpha. I hadn't thought about IF NOT EXISTS, but that's a good idea. also not in the alpha
</blockquote></details>

ah right, duh. hid for now

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/347b860e2367e8ac7405e8a4c114d9d38efec05a/

lhirata commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/create-changefeed.md, line 25 at r1](https://reviewable.io/reviews/cockroachdb/docs/3308#-LG6wQ8L4Vu4rOWmzwID:-LGGUeyY57BTkKTa2S7q:brnqdmv) ([raw file](https://github.com/cockroachdb/docs/blob/49eb216123131a53437ff7bad685ab12c1b56edd/v2.1/create-changefeed.md#L25)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

TO DO: generate new diagram
</blockquote></details>

will add actual diagram for 2.1

---

*[v2.1/create-changefeed.md, line 33 at r2](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGGUS-SBDjz1M4gpoMV:-LGGUS-SBDjz1M4gpoMW:bszlcno) ([raw file](https://github.com/cockroachdb/docs/blob/347b860e2367e8ac7405e8a4c114d9d38efec05a/v2.1/create-changefeed.md#L33)):*
> ```Markdown
> 
> Targets:
>     TABLE <table_name> [, ...]
> ```

Can we do changefeeds on partitions / individual rows yet? If so, how do we do that?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 21 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQTI1cEHyHCIbQxC9a:-LGQTI1cEHyHCIbQxC9b:bcc2wc9) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L21)):*
> ```Markdown
> While CockroachDB is an excellent system of record, it also needs to coexist with other systems. For example, you might want to keep your data mirrored in full-text indexes, analytics engines, or big data pipelines.
> 
> The core feature of CDC is the [`CHANGEFEED`](create-changefeed.html). Changefeeds target a whitelist of databases, tables, partitions, rows, or a combination of these; called the "watched rows." Every change to a watched row is emitted as a record in a configurable format (i.e., `JSON` or Avro) to a configurable sink (i.e., [Kafka](https://kafka.apache.org/).
> ```

should we remove references to partitions and rows here for the now?

---

*[v2.1/change-data-capture.md, line 135 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQVKQq5bC7Nje-TYMS:-LGQVKQq5bC7Nje-TYMT:bjdgmo7) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L135)):*
> ```Markdown
>     {% include copy-clipboard.html %}
>     ~~~ sql
>     > CREATE TABLE demo_cdc (
> ```

can we call this dogs and call the database cdc_demo? (or office_dogs or something)

---

*[v2.1/change-data-capture.md, line 197 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQRkeL-vqUCPOKjgUA:-LGQRkeL-vqUCPOKjgUB:b8q546l) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L197)):*
> ```Markdown
> ## Known limitations
> 
> <!-- TO DO: Dan to add -->
> ```

Whoops, this keeps slipping my mind. I'll do it right now.

---

*[v2.1/create-changefeed.md, line 33 at r2](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGGUS-SBDjz1M4gpoMV:-LGQRBgf1hriBx_tbFBY:b-dx8lzl) ([raw file](https://github.com/cockroachdb/docs/blob/347b860e2367e8ac7405e8a4c114d9d38efec05a/v2.1/create-changefeed.md#L33)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

Can we do changefeeds on partitions / individual rows yet? If so, how do we do that?
</blockquote></details>

nope. this is one of the alpha limitations

also, we can only do one table or database per changefeed right now (so no `[, ...]`)

---

*[v2.1/create-changefeed.md, line 36 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQSti2Arla5keGbCf3:-LGQSH768Zzs-VB35TJz:b-6u5abw) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/create-changefeed.md#L36)):*
> ```Markdown
>     DATABASE <database_name> [, ...]
> 
> Location:
> ```

Should this also be "sink"?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 197 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQRkeL-vqUCPOKjgUA:-LGQaPmI4tr45GBykSx7:b-a75klz) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L197)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Whoops, this keeps slipping my mind. I'll do it right now.
</blockquote></details>

Okay, here's what I have. The tense is all over the place, we should clean that up. Also, I tried to omit things like "does not yet", but that makes this list sound terrifying if you don't read the header. Thoughts?


The following are limitations in July 2nd alpha, but they will be addressed before the final 2.1 release.
- Changefeeds created with the alpha may not be compatible with future alphas and the final 2.1 release. *Do not use this feature on production data.* (Lauren: we should also add this to the big red warning we have at the top of the page.)
- The CockroachDB core CHANGEFEED is not ready for external testing.
- Some intermediate updates on frequently changed rows are never emitted. [#27101]
- Changefeed progress is not exposed to the user.
- The SQL interface is not final and may change.
- Changefeeds only work on tables with a single COLUMN FAMILY (which is the default for new tables).
- Changefeeds do not work on interleaved tables.
- Many DDL queries (including TRUNCATE, RENAME, and DROP) will cause undefined behavior on a CHANGEFEED watching the affected tables.
- Changefeeds cannot be backed up or restored.
- Changefeed behavior under most types of failures/degraded conditions is not yet tuned.
- Changefeed internal buffering does not respect memory use limitations.
- Changefeeds do not scale horizontally or to high traffic workloads.
- Changefeeds use a pull model, but will use a push model in 2.1, lowering latencies considerably.
- Changefeeds are slow on data recently loaded via RESTORE/IMPORT.
- Additional format options will be added, including Avro.
- Additional envelope options will be added, including one that includes the old and new values for the changed row.
- Additional target options will be added, including partitions and ranges of primary key rows.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Basically LGTM for today's release, after addressing some nits/questions and Dan's additional content. I can take another look once you're ready.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaG9l5xlKfIn0TnFC:-LGQaG9l5xlKfIn0TnFD:b4w69l7) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L7)):*
Given that this feature works only with Kafka, we probably want to mention that system in the intro. Here's what Nate wrote for the release notes, in case it helps:

> Stream changes to Kafka with CDC [enterprise] - CockroachDB can now stream changes into Apache Kafka to support downstream processing such as reporting, caching, or full-text indexing.

Or does this work with systems other than Kafka? If no, I think we might want to make sure the page doesn't read as though that's possible. This can be follow-up work though.

---

*[v2.1/change-data-capture.md, line 21 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQZXCK3JsiNBm-5m2d:-LGQZXCK3JsiNBm-5m2e:b1v0qtw) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L21)):*
> ```Markdown
> While CockroachDB is an excellent system of record, it also needs to coexist with other systems. For example, you might want to keep your data mirrored in full-text indexes, analytics engines, or big data pipelines.
> 
> The core feature of CDC is the [`CHANGEFEED`](create-changefeed.html). Changefeeds target a whitelist of databases, tables, partitions, rows, or a combination of these; called the "watched rows." Every change to a watched row is emitted as a record in a configurable format (i.e., `JSON` or Avro) to a configurable sink (i.e., [Kafka](https://kafka.apache.org/).
> ```

nits: 
- remove the semicolon in the first sentence.
- Add closing parenthesis at end of second sentence.

---

*[v2.1/change-data-capture.md, line 58 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQan4tBnhml0zrlGa0:-LGQan4tBnhml0zrlGa1:b-cwfxzj) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L58)):*
> ```Markdown
> 
> {% include copy-clipboard.html %}
> ~~~
> ```

Need to add `sql` to get syntax highlighting to work. 

---

*[v2.1/change-data-capture.md, line 64 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQbbLOAEn9Dm8NTroF:-LGQbbLOAEn9Dm8NTroG:bo7hibp) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L64)):*
> ```Markdown
> For more information, see [`PAUSE JOB`](pause-job.html).
> 
> ## Resume
> ```

Make this an h3.

---

*[v2.1/change-data-capture.md, line 69 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaqlX3Zf7nSdwEa7q:-LGQaqlX3Zf7nSdwEa7r:b-cwfxzj) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L69)):*
> ```Markdown
> 
> {% include copy-clipboard.html %}
> ~~~
> ```

Need to add `sql` to get syntax highlighting to work. 

---

*[v2.1/change-data-capture.md, line 80 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQarZy18eksDm6_vji:-LGQarZy18eksDm6_vjj:b-cwfxzj) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L80)):*
> ```Markdown
> 
> {% include copy-clipboard.html %}
> ~~~
> ```

Need to add `sql` to get syntax highlighting to work. 

---

*[v2.1/change-data-capture.md, line 90 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQbtzn8zDI5k4tKHGf:-LGQbtzn8zDI5k4tKHGg:b-wfphry) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L90)):*
> ```Markdown
> ### Create a changefeed connected to Kafka
> 
> In this example, you'll set up a changefeed that is connected to a Kafka sink.
> ```

I think this example is for testing CDC with a single-node local cluster? If so, we should be explicit about that. 

---

*[v2.1/change-data-capture.md, line 167 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQc4ZWD7hCZF0l2Gfl:-LGQc4ZXAOHl0FfWJwVy:b-h2rr8l) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L167)):*
> ```Markdown
>     ~~~
> 
>     This will start up the changefeed in the background and return the `job_id`. The changefeed writes to Kafka and survives node failures.
> ```

Since this is a single-node local cluster, node failures are kind of beside-the-point. I wonder if we should make this example a multi-node cluster, either local or real? We could then kill one node and prove that the changefeed keeps going. This can come later, though.

Also something for later: We might want to use one of the `workload` load gens to make this more interesting. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308#-:-LGQdKf7Bl7t8Z_bxu3A:bjbo84n)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 23 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQfUa6AEgVIVpFYs0e:-LGQfUa6AEgVIVpFYs0f:bqm7dxk) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L23)):*
> ```Markdown
> The core feature of CDC is the [`CHANGEFEED`](create-changefeed.html). Changefeeds target a whitelist of databases, tables, partitions, rows, or a combination of these; called the "watched rows." Every change to a watched row is emitted as a record in a configurable format (i.e., `JSON` or Avro) to a configurable sink (i.e., [Kafka](https://kafka.apache.org/).
> 
> ## Ordering guarantees
> ```

Here's a reordering of what you have in this section that I think is a bit clearer to me

- In the common case, each version of a row will be emitted once. However, some (infrequent) conditions will cause them to be repeated. This gives our changefeeds an **at-least-once delivery guarantee**.
- Once an update to a given row is emitted, it is guaranteed that all previous updates to that row have been emitted.
- If a row is modified more than once in the same transaction, only the last change will be emitted.
- Rows are sharded between Kafka partitions by the row’s [primary key](primary-key.html).
- The `WITH timestamps` option adds an **update timestamp** to each emitted row. It also causes periodic **resolved timestamp** messages to be emitted to each Kafka partition. A resolved timestamp is a guarantee that no (previously unseen) rows with a lower update timestamp will be emitted on that partition.
- Cross-row and cross-table order guarantees are not given. However, the resolved timestamp notifications on every Kafka partition can be used to provide strong ordering and global consistency guarantees by buffering records in between timestamp closures. (Because CockroachDB supports transactions that can affect any part of the cluster, there is no way to horizontally divide the cluster's transaction log in a way where each piece is independent, so it's impossible for us to scale this in the general case.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

danhhz commented:
I think that's everything I have, so LGTM to me after addressing outstanding comments (though I'd perfer to take a pass on the final version before merging)

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308#-:-LGQgdcT6CXOJoYWsd4g:b-e3qhwk)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/99f15f6291c3c208386edb46dbb1b070bc25a4c1/

lhirata commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[v2.1/change-data-capture.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaG9l5xlKfIn0TnFC:-LGQfNgU9wo72KQu8KOS:b8woslf) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L7)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Given that this feature works only with Kafka, we probably want to mention that system in the intro. Here's what Nate wrote for the release notes, in case it helps:

> Stream changes to Kafka with CDC [enterprise] - CockroachDB can now stream changes into Apache Kafka to support downstream processing such as reporting, caching, or full-text indexing.

Or does this work with systems other than Kafka? If no, I think we might want to make sure the page doesn't read as though that's possible. This can be follow-up work though.
</blockquote></details>

Updated. Only Kafka is supported - will update to be more specific.

---

*[v2.1/change-data-capture.md, line 21 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQTI1cEHyHCIbQxC9a:-LGQfVuG4YGt8ibuigpP:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L21)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

should we remove references to partitions and rows here for the now?
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 21 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQZXCK3JsiNBm-5m2d:-LGQfeCS6NYd7okjOAbV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L21)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nits: 
- remove the semicolon in the first sentence.
- Add closing parenthesis at end of second sentence.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 23 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQfUa6AEgVIVpFYs0e:-LGQkyXIBumy0mH9dp3E:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L23)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Here's a reordering of what you have in this section that I think is a bit clearer to me

- In the common case, each version of a row will be emitted once. However, some (infrequent) conditions will cause them to be repeated. This gives our changefeeds an **at-least-once delivery guarantee**.
- Once an update to a given row is emitted, it is guaranteed that all previous updates to that row have been emitted.
- If a row is modified more than once in the same transaction, only the last change will be emitted.
- Rows are sharded between Kafka partitions by the row’s [primary key](primary-key.html).
- The `WITH timestamps` option adds an **update timestamp** to each emitted row. It also causes periodic **resolved timestamp** messages to be emitted to each Kafka partition. A resolved timestamp is a guarantee that no (previously unseen) rows with a lower update timestamp will be emitted on that partition.
- Cross-row and cross-table order guarantees are not given. However, the resolved timestamp notifications on every Kafka partition can be used to provide strong ordering and global consistency guarantees by buffering records in between timestamp closures. (Because CockroachDB supports transactions that can affect any part of the cluster, there is no way to horizontally divide the cluster's transaction log in a way where each piece is independent, so it's impossible for us to scale this in the general case.)
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 58 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQan4tBnhml0zrlGa0:-LGQfpfCA5d0c_z3uLOQ:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L58)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Need to add `sql` to get syntax highlighting to work. 
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 64 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQbbLOAEn9Dm8NTroF:-LGQfr_0Cr_3ZgssJMdi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L64)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Make this an h3.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 69 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaqlX3Zf7nSdwEa7q:-LGQfs3D85tGrk5mcr4O:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L69)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Need to add `sql` to get syntax highlighting to work. 
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 80 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQarZy18eksDm6_vji:-LGQfsVC2z0DededElYt:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L80)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Need to add `sql` to get syntax highlighting to work. 
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 90 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQbtzn8zDI5k4tKHGf:-LGQg-vE3aBUzAOEAlIC:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L90)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

I think this example is for testing CDC with a single-node local cluster? If so, we should be explicit about that. 
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 135 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQVKQq5bC7Nje-TYMS:-LGQgJ8U5FI3nBOSXxUY:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L135)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

can we call this dogs and call the database cdc_demo? (or office_dogs or something)
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 167 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQc4ZWD7hCZF0l2Gfl:-LGQgQLzBR2wbty-gl2Z:btcfhqh) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L167)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Since this is a single-node local cluster, node failures are kind of beside-the-point. I wonder if we should make this example a multi-node cluster, either local or real? We could then kill one node and prove that the changefeed keeps going. This can come later, though.

Also something for later: We might want to use one of the `workload` load gens to make this more interesting. 
</blockquote></details>

Done. 

Will update the example later.

---

*[v2.1/change-data-capture.md, line 197 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQRkeL-vqUCPOKjgUA:-LGQicVLE6BZXWSk8trZ:bqmjwcz) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L197)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Okay, here's what I have. The tense is all over the place, we should clean that up. Also, I tried to omit things like "does not yet", but that makes this list sound terrifying if you don't read the header. Thoughts?


The following are limitations in July 2nd alpha, but they will be addressed before the final 2.1 release.
- Changefeeds created with the alpha may not be compatible with future alphas and the final 2.1 release. *Do not use this feature on production data.* (Lauren: we should also add this to the big red warning we have at the top of the page.)
- The CockroachDB core CHANGEFEED is not ready for external testing.
- Some intermediate updates on frequently changed rows are never emitted. [#27101]
- Changefeed progress is not exposed to the user.
- The SQL interface is not final and may change.
- Changefeeds only work on tables with a single COLUMN FAMILY (which is the default for new tables).
- Changefeeds do not work on interleaved tables.
- Many DDL queries (including TRUNCATE, RENAME, and DROP) will cause undefined behavior on a CHANGEFEED watching the affected tables.
- Changefeeds cannot be backed up or restored.
- Changefeed behavior under most types of failures/degraded conditions is not yet tuned.
- Changefeed internal buffering does not respect memory use limitations.
- Changefeeds do not scale horizontally or to high traffic workloads.
- Changefeeds use a pull model, but will use a push model in 2.1, lowering latencies considerably.
- Changefeeds are slow on data recently loaded via RESTORE/IMPORT.
- Additional format options will be added, including Avro.
- Additional envelope options will be added, including one that includes the old and new values for the changed row.
- Additional target options will be added, including partitions and ranges of primary key rows.

</blockquote></details>

Added.

---

*[v2.1/create-changefeed.md, line 33 at r2](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGGUS-SBDjz1M4gpoMV:-LGQitwM2YJV9yxfjkiX:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/347b860e2367e8ac7405e8a4c114d9d38efec05a/v2.1/create-changefeed.md#L33)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nope. this is one of the alpha limitations

also, we can only do one table or database per changefeed right now (so no `[, ...]`)
</blockquote></details>

Done.

---

*[v2.1/create-changefeed.md, line 36 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQSti2Arla5keGbCf3:-LGQjOiv5Zv70JDJrc1V:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/create-changefeed.md#L36)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Should this also be "sink"?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

lhirata commented:
@jseldess @danhhz TFTRs! Can you take a final look before I merge for the alpha? Thanks

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/93aa4a34a9032044a290aa5bcf5745c6423d4f58/

jseldess commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some final comments.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[v2.1/change-data-capture.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaG9l5xlKfIn0TnFC:-LGQsS774cP9uCiCZkUh:b-vyfyvv) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L7)):*
Thanks, Lauren. Would you allow me one more nitty rewrite?

> Change data capture (CDC) provides efficient, distributed, row-level change feeds into Apache Kafka for downstream processing such as reporting, caching, or full-text indexing.

Also, you need to add a space after `</span>`.

---

*[v2.1/change-data-capture.md, line 205 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQtfI90Rd9-uCYLTOo:-LGQtfIA-dS098ALQQpq:bchq7yi) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L205)):*
> ```Markdown
> 
> - The CockroachDB core `CHANGEFEED` is not ready for external testing.
> - Some intermediate updates on frequently changed rows are never emitted. [#27101]
> ```

Do we want `[#27101]` at the end of the line?

---

*[v2.1/change-data-capture.md, line 208 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuMv9EdVHK7Za12uj:-LGQuMv9EdVHK7Za12uk:b6slx2t) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L208)):*
> ```Markdown
> - Changefeed progress is not exposed to the user.
> - The SQL interface is not final and may change.
> - Changefeeds only work on tables with a single column family (which is the default for new tables).
> ```

Make `column family` a link to `column-families.html`.

---

*[v2.1/change-data-capture.md, line 209 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuW3aATVzJAUQfDO6:-LGQuW3aATVzJAUQfDO7:bb25jkv) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L209)):*
> ```Markdown
> - The SQL interface is not final and may change.
> - Changefeeds only work on tables with a single column family (which is the default for new tables).
> - Changefeeds do not work on interleaved tables.
> ```

Make `interleaved tables` a link `interleave-in-parent.html`.

---

*[v2.1/change-data-capture.md, line 211 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuaRS6o-7ec8NoPYX:-LGQuaRS6o-7ec8NoPYY:binul9h) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L211)):*
> ```Markdown
> - Changefeeds do not work on interleaved tables.
> - Many DDL queries (including `TRUNCATE`, `RENAME`, and `DROP`) will cause undefined behavior on a changefeed watching the affected tables.
> - Changefeeds cannot be backed up or restored.
> ```

Make `backed up` a link to `backup.html` and `restored` a link to `restore.html`.

---

*[v2.1/change-data-capture.md, line 216 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuhe9EEEcYfVpooPY:-LGQuhe9EEEcYfVpooPZ:b68hnec) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L216)):*
> ```Markdown
> - Changefeeds do not scale horizontally or to high traffic workloads.
> - Changefeeds use a pull model, but will use a push model in v2.1, lowering latencies considerably.
> - Changefeeds are slow on data recently loaded via `RESTORE` or `IMPORT`.
> ```

Make `RESTORE` a link to `restore.html` and `IMPORT` and link to `import.html`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308#-:-LGQutrZ6vsspQxfr4MB:b18u6sv)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[v2.1/change-data-capture.md, line 7 at r3](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQaG9l5xlKfIn0TnFC:-LGQvsCK6L_YtopV02Dk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/46f33721e208b948e98a999a5015ffaf75a9908e/v2.1/change-data-capture.md#L7)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Thanks, Lauren. Would you allow me one more nitty rewrite?

> Change data capture (CDC) provides efficient, distributed, row-level change feeds into Apache Kafka for downstream processing such as reporting, caching, or full-text indexing.

Also, you need to add a space after `</span>`.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 205 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQtfI90Rd9-uCYLTOo:-LGQw3vS5OpXpuOA2QaL:bss4a2b) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L205)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Do we want `[#27101]` at the end of the line?
</blockquote></details>

Removed.

---

*[v2.1/change-data-capture.md, line 208 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuMv9EdVHK7Za12uj:-LGQw8xR38qRTr_VoiVb:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L208)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Make `column family` a link to `column-families.html`.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 209 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuW3aATVzJAUQfDO6:-LGQwEH_3ED99De3mnCW:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L209)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Make `interleaved tables` a link `interleave-in-parent.html`.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 211 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuaRS6o-7ec8NoPYX:-LGQwKr88mkMpccdERk6:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L211)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Make `backed up` a link to `backup.html` and `restored` a link to `restore.html`.
</blockquote></details>

Done.

---

*[v2.1/change-data-capture.md, line 216 at r5](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQuhe9EEEcYfVpooPY:-LGQwR-E3iSAnoN0PhCW:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/93aa4a34a9032044a290aa5bcf5745c6423d4f58/v2.1/change-data-capture.md#L216)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Make `RESTORE` a link to `restore.html` and `IMPORT` and link to `import.html`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/03e6791aaf51c26a6bdabd826b9122a22971ae7e/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/589b9c1f418866bb178c41e27e31b0223c481c6b/

lhirata commented:
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---

*[v2.1/change-data-capture.md, line 35 at r6](https://reviewable.io/reviews/cockroachdb/docs/3308#-LGQzbqV1jrn7rvMes3V:-LGR1geC0cQD-RqckqFS:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/03e6791aaf51c26a6bdabd826b9122a22971ae7e/v2.1/change-data-capture.md#L35)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I realize I wrote this, but can you make this something like "are not directly given", plz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3308)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/1cbcade06a1bb0f069a24ce3640d0c51023441b2/
